Permissioned data (#94)
bluesky-social/proposals: 0016-permissioned-data/README.md

# 0016 Permissioned Data

> *This is a draft proposal, not the final specification. Details, terminology, and behaviors are all likely to change.*

Reviewer: This should link to the Discourse discussion thread (https://discourse.atprotocol.community/t/early-permissioned-data-proposal-draft-feedback/923/11).

I think it would also be appropriate to link to your leaflet "diary" posts as additional background context.

## Introduction

> The [AT Protocol][ATPROTO] is a protocol for public broadcast data. Users publish records into a repository on their PDS, and applications crawl those repositories to build views. Authority rests in the DID that publishes a record, and records are signed, redistributable, and universally addressable.

Reviewer: "AT Protocol is a framework which includes components for public broadcast data" or something like that? Emphasize that "atproto" is a big tent and includes multiple sync/transfer mechanisms on more-or-less equal footing. Would need to update in a couple places in this intro section.

This document specifies a **permissioned data protocol** for data that is _not_ public, data with an access perimeter. It runs alongside the public protocol and serves modalities such as:

- **Personal data**: bookmarks, mutes, drafts
- **Gated content**: paid newsletters, subscriber-only posts
- **Socially shared**: private posts, stories
- **Groups**: private forums, communities, group chats

The permissioned protocol shares the abstract shape of public atproto. It retains identity-based authority, per-user repositories, lexicon-typed records, and the general flow of applications crawling PDSes to build views. However it is a **distinct protocol** rather than an extension of the public one. It has its own repository format, sync mechanism, addressing scheme, and resolution path. Public atproto is built for public broadcast (signed, archival, rebroadcastable) while the permissioned protocol is built for party-to-party transmission within an access boundary.
 

> This protocol provides **access control, not confidentiality**. It is not end-to-end encrypted. Services (both PDSes and authorized applications) can read the data they handle, which is required for server-side features such as search, indexing, notifications, aggregation, and moderation. E2EE is a separate concern that may be layered on top by an application and is out of scope in this proposal.

Reviewer: have "not end-to-end encrypted" link to that specific blog post?

### Relationship to public atproto

| | Public atproto | Permissioned protocol |
|---|---|---|
| Unit of data | Record in a repo | Record in a permissioned repo |
| Repo scope | One repo per user | One permissioned repo per (user, space) |
| Record authority | User DID | User DID |
| URI authority | User DID | Space authority DID |
| Commit | Merkle Search Tree root | LtHash set-hash digest |
| Signature | Rebroadcastable, archival | Deniable on rebroadcast |
| Addressing | `at://` URI | `ats://` URI |
| Access | Public | Gated by space credential |

### Terminology

- **Space**: an authorization and sync boundary for a set of permissioned records, identified by an `(authority, type, skey)` triple.
- **Permissioned repo**: one user's records within one space, with a cryptographic commit, hosted on that user's PDS.
- **Repo host**: a service that stores and serves users' permissioned repos.
- **Space host**: a service that answers for a space as a whole, issuing credentials, enumerating writers, and routing notifications.

> - **Space authority**: the DID at the root of a space, which resolves to the space host and the key material for issuing credentials.

Reviewer 2: Thanks for eliminating the earlier "space owner" terminology.  Using "authority" here is much clearer

- **Space credential**: a token issued by the space authority that grants read access to a space.
- **Delegation token**: a token issued by a user's PDS that an application exchanges with a space authority for a space credential.

> - **Client attestation**: a token signed by an application's own client authentication key, proving the application's identity to a space authority. Required only when a space gates on app identity.

Reviewer: ```suggestion
- **Client attestation**: a token signed by an application's own client authentication key, proving the application's identity to a space authority. Required only when a space gates on client app identity.
```

- **Syncer**: an application that keeps its own copy of a space in sync by pulling from repo hosts.

A PDS fulfills both the roles of a **repo host** and a **space host**. However, these roles are discussed separately because they do not necessarily need to be filled by a PDS. A permissioned repo or a space may be hosted by any service that implements the [required APIs](#xrpc-api).

## Spaces

> A **space** is an authorization and sync boundary representing a shared social context. A space may include many different types of records from many users. The space does not colocate records. Instead, each user stores their own records for a given space in a [permissioned repo](#permissioned-repos) on their own repo host. A space is the union of these per-user repos across the network: an application presenting a space pulls each member's repo from its host, assembles the view, and applies access control to requesting users.

Reviewer: ```suggestion
A **space** is an authorization and sync boundary representing a shared social context. A space may include many different types of records from many users. The space does not collocate records in a single network host. Instead, each user stores their own records for a given space in a [permissioned repo](#permissioned-repos) on their own repo host. A space is the union of these per-user repos across the network: an application presenting a space pulls each member's repo from its host, assembles the view, and applies access control to requesting users.
```

> A **space** is an authorization and sync boundary representing a shared social context. A space may include many different types of records from many users. The space does not colocate records. Instead, each user stores their own records for a given space in a [permissioned repo](#permissioned-repos) on their own repo host. A space is the union of these per-user repos across the network: an application presenting a space pulls each member's repo from its host, assembles the view, and applies access control to requesting users.

Reviewer: ```suggestion
A **space** is an authorization and sync boundary representing a shared social context. A space may include many different types of records from many users. The space does not colocate records. Instead, each user stores their own records for a given space in a [permissioned repo](#permissioned-repos) on their own repo host. A space is the aggregation of these per-user repos across the network: an application presenting a space pulls each member's repo from its host, assembles the view, and applies access control to requesting users.
```

Each space is identified by three values:

> 1. **authority**: a DID, the root of authority for the space

Reviewer: ```suggestion
1. **space authority**: a DID, the root of authority for the space
```

> 2. **type**: an NSID describing the modality of the space

Reviewer: ```suggestion
2. **space type**: an NSID describing the modality of the space
```

> 3. **space key** (`skey`): a string distinguishing spaces of the same type under the same authority

Reviewer: ```suggestion
3. **space key** (`skey`): a string distinguishing spaces of the same space type under the same space authority
```

### Addressing

A permissioned record is addressed by an `ats://` URI of six components:

```
ats://{spaceDid}/{spaceType}/{skey}/{authorDid}/{collection}/{rkey}
```

| Component | Type | Description |
|---|---|---|
| `spaceDid` | DID | Space authority DID |
| `spaceType` | NSID | Space type |
| `skey` | string | Space key |
| `authorDid` | DID | DID of the record's author |
| `collection` | NSID | Record collection |
| `rkey` | string | Record key |

All six URI segments are necessary to identify a **permissioned record**. The first three components may be used to reference a **space**:

```
Space:  ats://{spaceDid}/{spaceType}/{skey}
Record: ats://{spaceDid}/{spaceType}/{skey}/{authorDid}/{collection}/{rkey}
```

### Space authority

A space's **authority** is the DID at the root of the space and the issuer of its [credentials](#access-control). It may be a user's own DID as for personal data such as bookmarks or mutes. Or it may be a dedicated DID which lets a shared space transfer between users independently of any individual account.

> The authority MUST expose two entries in its DID document:

Reviewer: "MUST": does this mean that every account in the network needs to update their DID document to make use of personal data?

I would have thought this would default to using the atproto signing key and PDS host as the equivalent space values if not specified. We'd be re-using the key for a similar type of use (just signatures, not encryption, with JWT domain separation) so I don't see cryptographic reuse concerns.

Author: No I'm not concerned about cryptograhpic re-use. However I am concerned about having a fallback in the specification. Just seems kinda footgun-y. Maybe it's not a bit deal. I agree that I'd strongly prefer not to require a migration for all DIDs.

Author: Okay decided to update this to allow fallback

> The authority DID MUST expose two entries in its DID document:

Reviewer 2: "The authority MUST expose ..."

Still clear, + preserves the capitalized MUST/MAY distinction without the potential misreading of a new weird past tense "DID MUST" state

- a **verification method** with id `#atproto_space`: the public key used to verify the space's credentials
- a **service** entry with id `#atproto_space_host`: the endpoint of the space host

Both values MAY resolve to the same values as `#atproto` and `#atproto_pds` from the public data protocol.

### Space type

> A space's **type** is an NSID that names its modality and resolves to a [space declaration](#space-type-declarations). It identifies the kind of data a space holds before any network resolution, much as a collection NSID does in public atproto. Because a type names a concrete modality, every space is some specific kind of space rather than a generic container.

Reviewer: ```suggestion
A space's **type** is an NSID that names its modality and resolves to a [space type declaration](#space-type-declarations). It identifies the kind of data a space holds before any network resolution, much as a collection NSID does in public atproto. Because a type names a concrete modality, every space is some specific kind of space rather than a generic container.
```

Reviewer: "rather than a generic container" could link to the diary blog post about that.

Author: I think I want to avoid diary/forum links throughout the proposal

The type is also the **OAuth consent boundary**. Access is granted to a user by type, e.g. "access to your AtmoBoards forums" (see [OAuth scopes](#oauth-scopes)).

### Space type declarations

> A space type NSID resolves to a **space declaration**: a Lexicon definition with `"type": "space"`.

Reviewer 3: it's wordy but could this always be written-out with "type" in there,

> A space type NSID resolves to a **space type declaration**: a Lexicon [...]

I slightly tripped on this being called just `space declaration` in some places, which sounds like it would be per-space concrete.

Author: Ah good point. Thanks 👍

```json
{
  "lexicon": 1,
  "id": "com.atmoboards.forum",
  "defs": {
    "main": {
      "type": "space",
      "description": "A discussion forum",
      "key": "any",
      "name": "AtmoBoards Forum",
      "name:lang": { "es": "Foro AtmoBoards", "ja": "AtmoBoards 掲示板" },
      "collections": ["com.atmoboards.thread", "com.atmoboards.reply"]
    }
  }
}
```

| Field | Type | Required | Description |
|---|---|---|---|
| `type` | `"space"` | yes | Marks this as a space declaration. Must be the `main` definition. |
| `description` | string | no | Description of the space type for developers. Not shown to users. |
| `key` | string | yes | Specifies the recommended space key type (similar to [record key types](https://atproto.com/specs/record-key#record-key-type-tid)) |
| `name` | string (1–64) | yes | Human-readable name for the space type, shown to users on consent screens. |

> | `name:lang` | map<lang, string> | no | Localized `name` values by language code. |

Reviewer 4: This is currently unrepresentable in Lexicon! It's something I've been fighting for for a while and experimenting with in [my own lexicons](https://tangled.org/tsunagite.dev/tsunagite/blob/main/schema/game.json#L138-159), and bnewbold has proposed a reified map type in [Expanding the Lexicon Language](https://bnewbold.leaflet.pub/3mimokt4o5c2f), but will need a Lexicon v2 to be possible to correctly represent and validate.

Reviewer 5: It is not a goal for lexicon documents to be able to schematize lexicon documents.

The `name:lang` here already exists, and is supported, in lexicon documents syntax, for permission sets.

Reviewer 4: oh, so it does! I got a bit mixed up between records and schema, whoops - a v2 of lexicon will still technically be necessary for defining a new type, but this will be representable.

## Access control

Reading a space is gated by a **space credential** issued by the space authority. The authority issues one based on two axes:

- **which user** is being acted for: established by a **delegation token** minted by the user's PDS
- **which application** is acting: established by a **client attestation** signed by the application itself

The delegation token is always required. The client attestation is required only when a space gates on app identity. An application obtains a credential by getting a delegation token from a user's PDS, then presenting that token (together with its client attestation, if needed) to the space authority in exchange for a credential. The authority decides whether to issue the credential. The protocol does not define the decision procedure (the policies of the PDS's space-management implementation are described under [`simplespace`](#required-pds-space-management-simplespace)).

> Some spaces do not require a client attestation. This by querying the space configuration, or by simply making a request for a credential without an attestation and seeing whether an error is returned.

Reviewer 2: "This can be detected by querying ..." 

(or however you wanna word this)

### Delegation token

> A **delegation token** proves that an application is acting on a user's behalf when it asks an authority for a space credential. The user's PDS mints it, requested through [`com.atproto.space.getDelegationToken`](#xrpc-api). It asserts only the delegation from user to application. Whether the user is a member of the space is the authority's determination, and the delegation token says nothing about it. The token is single-use, short-lived (default 60 seconds), and addressed (`aud`) to the space authority.

Reviewer: maybe worth mentioning here that client app sessions can only request delegation tokens for spaces which they have been authorized to access. eg, OAuth session scope, enforced by the PDS.

> A **delegation token** proves that an application is acting on a user's behalf when it asks an authority for a space credential. The user's PDS mints it, requested through [`com.atproto.space.getDelegationToken`](#xrpc-api). It asserts only the delegation from user to application. Whether the user is a member of the space is the authority's determination, and the delegation token says nothing about it. The token is single-use, short-lived (default 60 seconds), and addressed (`aud`) to the space authority.

Reviewer: single-use is noted here, but might want to make it a "MUST" that the space host enforce this using the nonce value.

Author: I think MUST might be a little aggressive & unnecessary on this one

Author: It requires extra implementation burden (for instance in our case some sort of cross-process cache) that I think a lot of space hosts don't really need

A delegation token is structurally similar to an atproto [service auth token](https://atproto.com/specs/xrpc), but it differs in a few ways that make it its own credential class rather than an interchangeable one: 
- The `typ` field in the header is set to `atproto-space-delegation+jwt`.
- It does not include an `lxm` claim.
- It is bound to a target space through the `sub` claim.

> Example JWT header and payload (before base64url encoding and signing):

Reviewer: It's a bit funky to me that this isn't (optionally) bound to `client_id`, if the client attestation is provided (eg, an OAuth session).

Author: Yeah that's how I originally had it, but @matthieusieben pushed back on it pretty hard over here: https://github.com/bluesky-social/proposals/pull/94#discussion_r3467426685

I ended up agreeing with his argument and dropping it. We don't action on it anywhere and we could always add it in if we need it in the future. For now, I think it's over-speccing

```json
{
  "typ": "atproto-space-delegation+jwt",
  "alg": "ES256K", // or ES256
  "kid": "#atproto" // Key Identifier - MUST be "#atproto"
}
{
  "iss": "did:example:user_did", // User DID

>   "aud": "did:example:space_did#atproto_space_host", // Space host (service fragment of the authority DID)

Reviewer 2: Consider flopping the aud/sub order to make for easier comparisons with the other two JWTs in this section (which both have sub immediately after iss)

  "sub": "ats://did:example:space_did/com.example.space_type/space_key", // Space being requested
  "iat": 1738368000, // Issued-at (unix seconds)

>   "exp": 1738368300, // iat + 60 (60 seconds)

Reviewer 2: Presumably exp should be 1738368060 to match the comment

  "jti": "f47ac10b58cc4372a5670e02b2c3d479" // random nonce
}
```

The delegation token asserts only the user-to-app delegation; it says nothing about which application is acting. App identity is established independently by the [client attestation](#client-attestation), which the authority verifies itself. The two are presented together but signed by different parties and evaluated independently.

The signature for the delegation token is computed using the regular JWT process, using the account's signing key. For more details, see the [Inter-Service Authentication](https://atproto.com/specs/xrpc#inter-service-authentication-jwt) section of the AT Protocol spec.

### Client attestation

> A **client attestation** is a short-lived JWT that the application presents to the space authority alongside the delegation token to identify itself, when the space requires it. It is structurally a `private_key_jwt` [client assertion](https://atproto.com/specs/oauth), the same shape an atproto confidential client already presents to its authorization server, but addressed to the space authority rather than to the PDS.

Reviewer: Clarify that client attestations are single-use (I assume), enforcable using the nonce value.

> A **client attestation** is a short-lived JWT that the application presents to the space authority alongside the delegation token to identify itself, when the space requires it. It is structurally a `private_key_jwt` [client assertion](https://atproto.com/specs/oauth), the same shape an atproto confidential client already presents to its authorization server, but addressed to the space authority rather than to the PDS.

Reviewer: A bit funky to me that this isn't bound/constrained to a specific space, only to a space host. Also not bound to an account DID.

> A **client attestation** is a short-lived JWT that the application presents to the space authority alongside the delegation token to identify itself, when the space requires it. It is structurally a `private_key_jwt` [client assertion](https://atproto.com/specs/oauth), the same shape an atproto confidential client already presents to its authorization server, but addressed to the space authority rather than to the PDS.

Reviewer: Want to think through non-OAuth client access. OAuth is the primary auth mechanism for now, but we do also have legacy auth clients/apps to deal with, and may also want to add support for "API tokens". Don't feel like we should make sync mechanisms too entangled/dependent on specific auth mechanisms.

- non-OAuth services and clients *could* still use a `client_id` an client metadata doc. it is some friction, but doable
- generating client attestations requires confidential clients; presumably this rules out local/dev (127.0.0.1) clients. I guess you can do most dev against "any client app allowed" spaces
- what if we allowed DIDs (or DID refs) in the `client_id`, combined with a service auth of some kind. this line of thinking leads towards something like client metadata docs, but as records in repos, either with DID or `at://` URI in the `client_id`

Author: This is for spaces that want to limit access to particular apps. Which I do think means that things like local/dev clients are out.

In cases like particular bot accounts (that use a token) or labelers (with no OAuth client) that you want to add to an app-limited space, I think you might want to allow those DIDs to bypass the client gate.

This is less of a protocol concern & more of a impl concern (for instance simplespace does not support something like that as currently specced).

```json
{
  "typ": "atproto-client-attestation+jwt",
  "alg": "ES256",
  "kid": "key-1" // Key id from the client's published JWKS
}
{
  "iss": "https://app.example.com/client-metadata.json", // The client_id
  "sub": "https://app.example.com/client-metadata.json", // The client_id (== iss)
  "aud": "did:example:space_did#atproto_space_host", // Space host being asked for a credential
  "iat": 1738368000, // Issued-at (unix seconds)
  "exp": 1738368060, // short-lived
  "jti": "b3c4d5e6f7a8b9c0d1e2f3a4b5c6d7e8" // random nonce, replay protection
}
```

The authority verifies the attestation by resolving `iss` (the `client_id`) to the client's `client-metadata.json`, fetching its published JWKS (`jwks` or `jwks_uri`), and verifying the signature against the key in that JWKS identified by the attestation's `kid`. 

### Space credential

> A **space credential** is the token an application presents to a repo host to read a permissioned repo within a space. The authority mints it in exchange for a delegation token, requested through [`com.atproto.space.getSpaceCredential`](#xrpc-api). It is short-lived (default 2 hours) and signed by the space authority's signing key, so any repo host can verify it against the authority's key without contacting the authority.

Reviewer: As I understand it, the space credential is multi-use per repo host, and can be reused across many repo hosts. Repo hosts would not enforce single-use via the nonce field. Probably worth clarifying that here.

A space credential resembles a [space delegation token](#delegation-token), differing in these ways:

- The `typ` field in the header is set to `atproto-space-credential+jwt`.
- It is signed by the space authority rather than the user.
- It has no `aud`: it is presented to any repo host serving a repo in the space, not to a single recipient.

> Example JWT header and payload (before base64url encoding and signing):

Reviewer: I'm not certain whether the space credential should include the original reading account's DID or not. Including it would be a bit muddy (the app might be providing the requested and indexed data to many users, not just the one), and could have privacy impacts (repo hosts could log that info to know who-saw-what). It make it possible for repo hosts to block specific readers faster than the token expiration window, though how the repo host would know to want to do that wouldn't be specified.

Author: I originally specified it that way, but then removed it because I felt the privacy hit was too harsh. And it makes the story for applications that juggle credentials for multiple users even fuzzier - it now possibly matters which credential you choose

```json
{
  "typ": "atproto-space-credential+jwt",
  "alg": "ES256K", // or ES256
  "kid": "#atproto_space" // Key Identifier - MUST be "#atproto_space"
}
{
  "iss": "did:example:space_did", // Space authority DID
  "sub": "ats://did:example:space_did/com.example.space_type/space_key", // Space the credential reads

>   "client_id": "https://app.example.com/client-metadata.json", // Attested application (the verified attestation's iss; omitted if there was none)

Reviewer 5: the `"client_id"` field in the **Space credential** has several issues:

1) In small groups it allows member to easily infer which user uses which app (privacy)
2) Some clients use query parameters (e.g. loopback client) to put potentially sensitive data (privacy/tracking/security)

I strongly recommend that value **not** be sent to the Repo Host through the space credential.

Author: Okay I think you're right. The reason to include it would be to allow a repo host to make the decision not to serve content to some app. But I think I agree that that is the wrong way to make that determination. 

Removed for now 👍

The signature for the space credential is computed using the regular JWT process, using the space authority's signing key.

### Credential flow

```
┌──────┐          ┌────────────┐          ┌─────────────┐          ┌─────────────────┐
│ User │          │ User's PDS │          │ Application │          │ Space Authority │
└───┬──┘          └──────┬─────┘          └──────┬──────┘          └────────┬────────┘
    │                    │                       │                          │
    ├── OAuth consent ───►                       │                          │
    │                    ├───── OAuth token ─────►                          │
    │                    │                       │                          │
    │                    ◄─ getDelegationToken ──┤                          │
    │                    ├── delegation token ───►                          │
    │                    │                       │   getSpaceCredential     │
    │                    │                       ├─(token [+ attestation])─►│
    │                    │                       ◄──── space credential ────┤

> ```

Reviewer 2: This is an excellent diagram.   If you have time down the road, might be nice to have something similar to clarify how sync flows work to push changes downstream.  Might help clarify the roles of repo host vs. space host vs. syncer vs. apps

But this can wait until the design firms up

1. The user authorizes the application via OAuth.
2. The application calls `com.atproto.space.getDelegationToken` on the user's PDS, receiving a delegation token.
3. The application presents the delegation token to the space authority via `com.atproto.space.getSpaceCredential`, adding its own client attestation if the space gates on app identity. The authority verifies what it received and, on authorization, returns a space credential.
4. The application reads the repo from each member's repo host with the credential.

An application serving several users of a space does not necessarily need to maintain a space credential for each user. It may obtain its credential using any one user's session. When it loses all OAuth sessions for a space, it can no longer renew the credential and loses access.

## Permissioned repos

> A **permissioned repo** is one user's set of records within one space, stored on that user's own PDS. A user has one permissioned repo per space it participates in. Abstractly, a permissioned repo offers a similar interface to a public [atproto repository](https://atproto.com/specs/repository): a key/value mapping where keys are path names (a collection NSID and an rkey) and values are CBOR-encoded records. PDSes expose a CRUD interface for interacting with a user's permissioned repos.

Reviewer 6: Minor inconsistency, text on line 45 seems to un-bundle the user's PDS and their permissioned repos, this seems to bundle them

Author: loosened that language a bit 👍

### Commit digest

The digest in a commit is computed from a **set hash** over the records a repo currently contains. It is independent of write/delete order, so two repos with the same records always produce the same digest. Adding or removing a record is a single cheap operation, so a repo host maintains the set hash incrementally on each write.

The construction is [LtHash][LTHASH], a homomorphic hash built on a lattice problem (and thus quantum-secure). A record is added to or removed from the digest with a single addition or subtraction rather than a recomputation over the whole repo.

> The state is a fixed 2048-byte buffer, interpreted as 1024 little-endian unsigned 16-bit lanes.

Reviewer: will be interesting to see the database/disk write churn impact of this for high-update-rate spaces. eg sqlite and/or postgresql WAL size, VACUUM times. both for PDS hosts and syncers. doesn't all show up as bandwidth, but does show up as I/O.

The state is a fixed 2048-byte buffer, interpreted as 1024 little-endian unsigned 16-bit lanes.

> Each record maps to one element, the UTF-8 bytes of `{collection}/{rkey}/{record_cid}`. As with the public repository, the components are currently limited to ASCII, so the encoding (though specified as UTF-8) is a no-op.

Reviewer 2: "Each record in the repo maps to one element ..."  (reads slightly clearer to me)

Kudos for the elegance of your three-part key here, BTW.  It's really nice to trade the first four (redundant) segments of the fully-qualified space URI in favor of the CID.  Nicely done!

To **add** an element:

1. Expand the element to 2048 bytes with BLAKE3 in XOF (extendable-output) mode.
2. Read those bytes as 1024 little-endian `uint16` lanes.
3. Add each lane into the corresponding state lane, modulo 65536 (2^16, i.e. with wraparound)

To **remove** an element, perform the same process, but subtract its lanes instead (modulo 65536). 

Both operations are commutative, so the state depends only on the current set of records, not the order of writes. The empty repo's state is all zeroes.

> The commit's `hash` is `sha256(state)`, a 32-byte digest of the 2048-byte buffer. The repo host maintains the full state. Only the `hash` is carried in a commit.

Reviewer 2: In a later phase, I assume there'll be a security analysis discussing why collision resistance for a digest of this size isn't a practical issue here

Y'know, for those of us who aren't experienced enough to do this in our heads

Author: Yeah I think a security considerations section would be good. basically: sha256 with 32-bytes is considered collision-resistant.

A user does not sign the digest directly since a signature over the content digest would be a rebroadcastable proof of what the user wrote in a private space. Instead the signature covers only random per-commit bytes, and the digest is bound to those bytes by a symmetric MAC. A reader in the sync flow gets full authenticity and integrity, but a leaked commit is deniable and proves nothing about its contents to a third party.

Both the signature and the MAC are domain-separated by a single context string, `ctx`, built once and reused for both. `ctx` uses the variable-length-vector encoding from [TLS 1.3][RFC8446] (§3.4). It is composed of a fixed protocol tag followed by each variable field length-prefixed with a big-endian `uint16`.

Note: these length prefixes are big-endian, following the TLS convention for wire encodings. This is the opposite byte order from the little-endian lanes of the [commit digest](#commit-digest), which follows the LtHash reference construction. The two come from different specs and each keeps its native byte order.

```

> ctx = "atproto-space-v1"            // fixed protocol tag

Reviewer: should the author DID be in here too? thinking about a corner case where somebody tries to replay a signed ctx with the same ikm and repository state, but from a separate account/DID. not sure what that replay would gain anybody.

Author: Yeah probably, it can't hurt anyway

   || uint16be(len(space)) || space  // space URI (ats://authority/type/skey)
   || uint16be(len(rev))   || rev    // commit revision (TID)

>    || uint16be(len(ikm))   || ikm    // per-commit nonce (below)

Reviewer: "per-request nonce"?

Author: per-signature?

```

A commit is then produced as follows:

1. Generate `ikm`, 32 fresh random bytes. A new `ikm` is generated for each reader the commit is served to.
2. Compute `sig = sign(ctx)` with the user's signing key. The signed message only contains the space, revision and ikm, not the current repository hash. 
3. Compute `mac = HMAC-SHA256(HKDF-SHA256(ikm, ctx), hash)`, binding the repository hash to this commit's context.

> A reader verifies `sig` against the user's signing key (authenticity), then recomputes `mac` and compares (integrity). Because the digest is bound by a *symmetric* MAC keyed from the public `ikm`, anyone holding the commit can compute a valid `mac` for any `hash`, so a rebroadcast commit cannot prove what the user wrote, only that they signed a `(space, rev, ikm)` context. 

Reviewer: note that this does constitute cryptographic proof that the account contributed (wrote) to the given space (or at least attempted to write in to the space)

> The signed commit (`com.atproto.space.defs#signedCommit`):

Reviewer: we've got domain separation above (good), but I think this object should contain a version field (`ver` or `version`, integer type, starting with 1)

| Field | Type | Description |
|---|---|---|
| `hash` | bytes | `sha256` of the LtHash state (32 bytes) |
| `ikm` | bytes | per-commit nonce (32 random bytes) |
| `sig` | bytes | `sign(ctx)` by the user's signing key |
| `mac` | bytes | `HMAC-SHA256(HKDF-SHA256(ikm, ctx), hash)` |
| `rev` | string | commit revision (TID), also bound into `ctx` |

## Sync

Permissioned data sync is functionally similar to public atproto. Applications build views by pulling repos from their hosts. The major difference is that there is no relay to provide a collated firehose of data for the network as permissioned repositories are by their nature non-rebroadcastable. An application pulls directly from each repo host and is responsible for keeping its own copy in sync.

### Incremental sync

A syncer keeps its own copy of a repo and, alongside it, its own running [set hash](#commit-digest) over that copy. This is the same digest that the repo host maintains. When the two digests agree, the syncer knows that their copy is exactly up-to-date with the hosted repo. If there is a disagreement, then the syncer knows their copy is behind or corrupted.

> To advance, a syncer calls `com.atproto.space.listRepoOps` with a `since` revision. A repo host keeps an **operation log** of recent writes to each repo and returns the operations after `since`. Each entry is `{ rev (TID), collection, rkey, cid, prev }`. The field `cid` is null for a delete and `prev` is null for a create. Multiple records may be mutated atomically, and this is captured by entries sharing a `rev`. The syncer applies each operation to its copy and updates its running set hash accordingly.

Reviewer: does `listRepoOps` include the record value (data) itself, or does `getRecord` need to be called as well?

I assume we want a way to reduce the number of XRPC API calls, so there should be a way to get the record data along with the ops metadata. In that case, if incrementally fetching `since` multiple commits, any "stale" (since deleted/updated) record data should not be included.

Author: Yeah I think you're right this should return it by default

If a given oplog response includes the last available operation that occurred to a repo, then the response must also include the member's current signed [commit](#commit-signature). The syncer compares the commit's `hash` against its own running set hash. If they match, the syncer is fully in sync and the commit's signature authenticates that state. If they do not, the syncer has diverged and must fall back to a recovery process.

Sync is self-healing because its correctness rests on the set hash comparison rather than on receiving every operation. A missed operation is detected on a subsequent sync. A total disjunction in repository history is also recognizable, and is recovered from by syncing the repository contents in their entirety.

> The oplog is a transport optimization rather than a committed data structure. Its contents and history are not guaranteed, and a repo host may compact or drop it, retaining only a backfill window.

Reviewer: Worth mentioning that it would be reset on account migration.

### Full-state recovery

In cases in which a syncer cannot proceed incrementally, it must recover by syncing the full state of the repository.

> To do so, a syncer enumerates the repo's structure (paths → CIDs) with `com.atproto.space.listRecords`, diffs that against its local copy, and fetches the record values it is missing with `com.atproto.space.getRecord`. It then rebuilds its set hash from the recovered structure and compares it against the signed commit. 

Reviewer: similar question to above: seems like it would be good to have a version of `listRecords` which includes the record data itself (for full backfill, avoiding many `getRecord` calls), and a version with just the metadata (for "healing" and a smaller number of selective `getRecord` calls)

To do so, a syncer enumerates the repo's structure (paths → CIDs) with `com.atproto.space.listRecords`, diffs that against its local copy, and fetches the record values it is missing with `com.atproto.space.getRecord`. It then rebuilds its set hash from the recovered structure and compares it against the signed commit. 

### Blob sync

Blobs referenced by permissioned records are stored on the authoring repo's host and fetched via `com.atproto.space.getBlob` with the space credential. 

### Write notifications

Write notifications inform syncers that a repo has advanced, so they can pull promptly instead of continuously polling for updates. A notification contains no record data, and states only that a given repo in a given space has reached a new revision.

> A syncer subscribes to notifications by calling `com.atproto.space.registerNotify`. When called on a space host, this method subscribes to writes for all repos in a space. It can also be called on particular repo hosts to receive notifications for specific repos. `registerNotify` is authenticated with a space credential. The service that was registered against should return the expiration time for the registration which may be longer than the expiration window of the space credential.

Reviewer 2: "When called on a space host... " vs. "called on particular repo hosts ..." 

There's an architectural expectation in the public protocol that relays + PDS-level aggregation help reduce fanout pressure on individual repos.  I keep looking for similar rollup points in this protocol.  Should there be stronger advice that it's preferable to concentrate updates to a space via syncers +/or space hosts?  Or is it considered fine for every reader to register with every repo host whose currently writing to that space?

Author: good clarification 👍 

yes generally notifications get fanned out via the space host which is in a better position to track all recipients. If a syncer wants redundancy, they're free to subscribe to every repo host as well. Since these are such lightweight notifications, I expect it to be okay

> When a member writes, their PDS sends a `com.atproto.space.notifyWrite` to each endpoint registered for that repo. A PDS may not otherwise know which services are syncing the space, which is why the **space authority** registers itself as a subscriber on each repo host. Members notify the authority, and the authority forwards each notification to the endpoints registered with it for the space. Each notified syncer then pulls the updated repo directly from the relevant repo host. The authority only routes notifications and does not carry record data.

Reviewer: does the space authority need to explicitly register itself with each repo host? how would it know when it needs to do that? I'd expect repo hosts to auto-register the space authority's host for each non-personal-data repo upon first write. and the space authority could then un-register if it wanted.

don't need to get in to it here, but we may end up needing ways to deal with request rates and limits. eg, batching notifications, linger behaviors, circuit breaking, backoffs.

> Notifications are **best-effort** and are not required for correctness. If a notification is dropped, the affected repo is caught up by a later write's notification, or by a periodic sweep by the syncer.

Reviewer: you might already have it in your branch, but we probably need a `getHead`/`getCommit` sort of endpoint for these periodic sweeps. maybe even a `listSpaceCommits` which enumerates all repos participating in a space on the current repo host?

Author: what about just moving that into `listRepos` a la `com.atproto.repo.listRepos?`

Reviewer: that would definitely be more efficient, but what would the auth story be for that against repo hosts? space credentials are bound to a single space.

the space authority/host could support an endpoint like this, but it wouldn't be authoritative (eg, wouldn't catch missed notifications from repo host to space host).

### The sync boundary (writer set)

> Syncing a single repo requires the syncer to know that the repo exists. To sync a space in full, or to begin syncing a space for the first time, an application requires the set of accounts whose repos hold data in the space. This **writer set** is retrieved from the space authority via `com.atproto.space.listRepos`.

Reviewer: ```suggestion
Syncing an individual repo requires the syncer to know that the repo exists. To sync a space in full, or to begin syncing a space for the first time, an application requires the set of accounts whose repos hold data in the space. This **writer set** is retrieved from the space authority via `com.atproto.space.listRepos`.
```

Because it subscribes to updates from every repository in the space, the authority can easily maintain a complete and current record of which repos hold data in the space as well as their current rev.

> The writer set is a simple fetch and carries no commit or history. It enumerates only accounts that have **written** to the space. Accounts that may read the space are never enumerated.

Reviewer: Should this return the set of all accounts which are allowed to write in to the space, or only the set which have created a space repo on their PDS (which might be empty), or who have written at least one record in to the space?

Also might be worth noting that this list is what is claimed by the space authority (host). Which isn't authoritative for whether the account has actually "joined" or contributed to the space; need to do the sync request to the repo host to determine that.

Author: Yeah only those that have written at least one record in the space.

> The writer set is a simple fetch and carries no commit or history. It enumerates only accounts that have **written** to the space. Accounts that may read the space are never enumerated.

Reviewer 2: "... are never enumerated."

This may become a UX problem.  It's hard for users to give informed consent on whether they want to share stuff if they don't have a trusted way to know who'll see it.  Further down, the simplespace baseline does provide a listMembers endpoint, but if this isn't a common endpoint that other implementations must support, it's hard to see how to set baseline UX expectations here

Is the intent to gather more requirements + maybe standardize later, or to just stay out of the business entirely?

Author: The intent is to handle that at a different layer. Most likely via records in the space authority's repo or via an implementation-specific endpoint like `listMembers`. Ultimately in any of these cases, a service can always lie. Even an an earlier design where the member list was represented at the protocol level, a space authority could still issue credentials to other DIDs not on the list.

A space may be deleted by its authority. The authority stops issuing credentials and no longer answers for the space. The authority also deletes its own repo in the space. 

> The authority then notifies registered syncers and known repo hosts of participating members with `com.atproto.space.notifySpaceDeleted`, over the same best-effort path as write notifications. 

Reviewer: I wonder if we will want all space notifications to go to a single "hot" endpoint instead of breaking them out by message type. From an operational standpoint, rate-limiting, etc.

Message passing!

Author: Yeah deletion seems different enough (it's a heavier operation & it's more important to not drop) that I figured it was worthwhile to give it its own method

The authority then notifies registered syncers and known repo hosts of participating members with `com.atproto.space.notifySpaceDeleted`, over the same best-effort path as write notifications. 

> A **syncer** should delete every copy of the space's data it holds, both the repos it pulled and any derived state, as it is no longer authorized to retain them. A **repo host** should instead flag the member's repo as belonging to a deleted space rather than erase it. The data is the user's own, so the host decides how and whether to surface it to the user (e.g. an export, a grace period) before garbage-collecting.

Reviewer 2: "... so the repo host decides how and whether ..."

Clarification.  Had to read this twice to confirm that the space host wasn't making that policy decision

A user grants an application access to spaces through a `space:` OAuth scope. The scope identifies a set of spaces by their `(did, spaceType, skey)` identifier and states what the grant permits within them.

```
space:<spaceType>[?did=<did>][&skey=<skey>][&collection=<nsid>...][&action=<action>...][&manage=<op>...]
```

| Parameter | Position | Multiple | Default | Values |
|---|---|---|---|---|
| `spaceType` | positional (required) | no | N/A | a space-type NSID, or `*` for any type |

> | `did` | query | no | `*` | a space authority DID, or `*` for any authority |

Reviewer: kind of bike-shedding, but on first read I think `did` is a bit confusing here? `authority` or `spaceDid`? might just need to get comfy with it.

I think there should probably be a `self` special value for personal data. and maybe even have that be the default. I sorta think that personal data (prefs etc) is going to be the most common use of spaces, even if they don't get the most attention.

Author: yeah i like `authority`

and i think I like the addition of `self` as well. it's nice that that allows for static (but still limited) scopes

| `manage` | query | yes | _(none)_ | `create`, `update`, `delete` |

`did`, `spaceType`, and `skey` select **which spaces** the grant covers, matching the first three segments of a space URI. `action` (and `collection`) govern operations on the **records** in those spaces. `manage` governs operations on the **spaces themselves**.

### Read access

`read` is all-or-nothing at the space boundary. There is no partial, per-record, or per-author whole-space read grant.

A `read` grant confers two things:

- access to the read and sync methods (`com.atproto.space.getRecord`, `listRecords`, `getBlob`, `getRepoState`, `listRepoOps`) on the holder's own PDS, sufficient to read the holder's own repo in the space
- access to `com.atproto.space.getDelegationToken` for that space, which an application exchanges for a space credential to read  any repo in the space

`read_self` is the narrower grant. It confers access to the same read and sync methods, but only for the holder's **own** repo in the space, and it does **not** grant `getDelegationToken`. An application holding only `read_self` can read its user's own records but cannot reach the rest of the space. `read` implies `read_self`.

> A **space credential** grants whole-space read/sync access directly, so the read and sync methods accept **either** a covering OAuth scope or a space credential. Write methods accept only an OAuth credential, since a write is attributed to the authoring user.

Reviewer: "only an OAuth credential"... or legacy session auth, or future API token auth?

Author: I kinda think legacy session auth just shouldn't allow these. And API tokens will presumably re-use OAuth scopes. I'd just leave this as-is for now imo

Reviewer: Agree on re-using OAuth scope system for future API tokens.

Enthusiastic if this all ends up being a forcing function to get more apps over on to OAuth.

A bit bummed if `goat` and simple headless tools (CI, bots, scripts) can't work with permissioned spaces w/o switching to OAuth.

A **space credential** grants whole-space read/sync access directly, so the read and sync methods accept **either** a covering OAuth scope or a space credential. Write methods accept only an OAuth credential, since a write is attributed to the authoring user.

### Matching

A request is authorized by a grant when its target space matches the grant's `(did, spaceType, skey)` (each component equal to the grant's value or covered by its `*`) and the grant permits the requested operation, per the rules below.

**Record operations** are governed by `action`:

> - `read` covers every repo in the space and ignores `collection`.

Reviewer 2: "... and ignores ..."  

After reading the whole spec, I don't necessarily disagree, but this was surprising on first read-through, because it wasn't immediately clear why.   Might just be me though, so see if others trip on this before making any changes here

Author: I mention the all-or-nothing nature in the prev section. But I added a quick reiteration here as well 👍

Omitting `action` grants `read`, `create`, `update`, and `delete` (`read` is inclusive of `read_self`).

> Omitting `collection` defaults it to the `collections` declared by the space type's [declaration](#space-type-declarations). A bare `space:com.atmoboards.forum` grant therefore permits writing the forum's own record types, the same way a bare `repo:` scope permits writing the collections it names. An application may narrow this by listing a subset of collections, or widen it with `collection=*`. When `spaceType` is `*` there is no declaration to draw from, so the default is empty and the grant confers no write targets unless provided.

Reviewer: not mentioned here, but I assume the semantics around updates to the space type declaration lexicon schema are similar to those around updating permission set schemas? eg, that adding collections to this array would dynamically update authz grants for existing sessions, within some cache/rollout time window.

Author: yup. I figured it wasn't necessary to reiterate

| `putRecord` | pds | procedure | OAuth | Create or update a record. |
| `deleteRecord` | pds | procedure | OAuth | Delete a record. |
| `applyWrites` | pds | procedure | OAuth | Apply a batch of creates, updates, and deletes to one repo atomically. |
| `listSpaces` | pds | query | OAuth | The spaces the caller holds a repo in. |
| `registerNotify` | repo/host | procedure | space credential | Register an endpoint to be notified of writes. On the space host, subscribes to the whole space. On a repo host with a `repo`, subscribes to that repo. |
| `notifyWrite` | syncer/host | procedure | service auth | Notify that a repo advanced. Sent by a repo host to the space host, and forwarded by the space host to registered syncers. |
| `notifySpaceDeleted` | syncer/repo | procedure | service auth | Notify that a space was deleted. Sent by the authority to members and registered syncers. |

## Required PDS space management: `simplespace`

The protocol does not specify how spaces are created or how an authority decides who may read them. Those are the concern of each space-management implementation, which sits above the protocol and is identified by its own lexicon namespace.

> `com.atproto.simplespace` is the space-management implementation that every PDS MUST support. It gives applications a baseline that is available on every account's PDS to build against. `simplespace` spaces are anchored on a user's own DID and governed by an explicit member list (or the `public` and `managing-app` describe policies below).

Reviewer: ```suggestion
`com.atproto.simplespace` is the space-management implementation that every PDS MUST support. It gives applications a baseline that is available on every account's PDS to build against. `simplespace` spaces are anchored on a user's own DID and governed by an explicit member list (or the `public` and `managing-app` policies described below).
```

`com.atproto.simplespace` is the space-management implementation that every PDS MUST support. It gives applications a baseline that is available on every account's PDS to build against. `simplespace` spaces are anchored on a user's own DID and governed by an explicit member list (or the `public` and `managing-app` describe policies below).

`simplespace` is neither the only permitted implementation nor a privileged one. It is simply the one that PDSs are required to support. Other space types may define their own management implementations and are full protocol participants, but they are hosted on bespoke space services rather than on the PDS.

> | Method | Type | Description |

Reviewer 2: Presumably a future drafts will include an additional column describing the Auth patterns needed here

My initial guess = whatever "OAuth/space credential" meant in the previous table

Author: These actually all just require an OAuth credential. I added a line before the table 👍

|---|---|---|
| `createSpace` | procedure | Create a space (caller becomes the authority) |
| `updateSpace` | procedure | Update config (more details below) |
| `deleteSpace` | procedure | Delete the space (see [Space deletion](#space-deletion)). |
| `addMember` | procedure | Add a member (by DID) to view the space. |
| `removeMember` | procedure | Remove a member (by DID). |
| `listMembers` | query | List the current members of a space. |

### Configuration

`simplespace` spaces can be further configured along a few dimensions. This configuration is updated through `updateSpace` and is surfaced through `com.atproto.space.getSpace`.

| Field | Values | Description |
|---|---|---|

> | `mintPolicy` | `public` \| `member-list` \| `managing-app` | How the authority decides whether to authorize a _user_. |

Reviewer: "mint" feels a bit weird/jargon-y to me. What about `memberPolicy`?

Author: simplified to just `policy`

- `public`: authorize any requester.
- `managing-app`: at mint time, ask `managingApp` whether to authorize the request, via [`checkUserAccess`](#the-managing-app) below. Enables dynamic policies (e.g. follower-gating) without an app maintaining a list.

**App access** decides per-app authorization. It is an open union with two current variants:
- `#open` (default): any application may access the space. No [client attestation](#client-attestation) is required, so public clients work.
- `#allowList`: only the apps named in `allowed` may access the space. The authority evaluates the list against the **attested** `client_id` (the `iss` of the verified client attestation), so it is enforceable rather than advisory.

### The managing app

Spaces may list a `managingApp`. Any account with access to the space can read the selected managing app in the space config through `com.atproto.space.getSpace`. This endpoint may be used to route certain application-level requests that cannot be handled by a generic PDS implementation, such as join requests. Calls to managing apps are generally authenticated using service auth tokens.

> When a space's `mintPolicy` is `managing-app`, the space authority defers to the space's `managingApp` at mint time by calling `com.atproto.simplespace.checkUserAccess`. 

Reviewer 2: Aha.  In really complex scenarios -- where the space host + managingApp are implemented by the same service -- this wouldn't be needed, but putting an API boundary between them feels right-ish to me too.   Looking forward to more details + scenarios as you flesh this out more
